File: apps/windows/ReviewFault/Services/SyncClient.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using ReviewFault.Data;

namespace ReviewFault.Services;

public sealed record AuthSession(
    string AccountId, string WorkspaceId, string AccessToken,
    long AccessExpiresAt, string RefreshToken);
public sealed record PullResult(long Cursor, IReadOnlyList<PulledOperation> Operations);

public sealed class SyncClient : IDisposable
{
    private readonly HttpClient http;
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    public SyncClient(string endpoint)
    {
        var normalized = endpoint.Trim().TrimEnd('/');
        if (!Uri.TryCreate(normalized, UriKind.Absolute, out var uri) ||
            !string.IsNullOrEmpty(uri.UserInfo) || !string.IsNullOrEmpty(uri.Query) ||
            !string.IsNullOrEmpty(uri.Fragment) || uri.AbsolutePath != "/" ||
            (uri.Scheme != Uri.UriSchemeHttps && !(uri.Scheme == Uri.UriSchemeHttp &&
              (uri.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase) ||
               uri.Host == "127.0.0.1" || uri.Host == "[::1]"))))
            throw new ArgumentException("同步地址必须使用 HTTPS（localhost 除外）");
        http = new HttpClient { BaseAddress = uri, Timeout = TimeSpan.FromSeconds(30) };
    }

    public async Task RegisterAsync(string email, string password, string invitationCode) =>
        await SendAsync(HttpMethod.Post, "/api/v1/auth/register",
            new { email = email.Trim(), password, invitationCode = invitationCode.Trim() }, null,
            [System.Net.HttpStatusCode.Accepted]);

    public async Task<AuthSession> LoginAsync(
        string email, string password, string deviceId, string deviceName) =>
        Session(await SendAsync(HttpMethod.Post, "/api/v1/auth/login",
            new { email = email.Trim(), password, deviceId, deviceName }));

    public async Task<AuthSession> RefreshAsync(string deviceId, string refreshToken) =>
        Session(await SendAsync(HttpMethod.Post, "/api/v1/auth/refresh",
            new { deviceId, refreshToken }));

    public async Task LogoutAsync(string accessToken) =>
        await SendAsync(HttpMethod.Post, "/api/v1/auth/logout", new { }, accessToken,
            [System.Net.HttpStatusCode.NoContent]);

    public async Task<IReadOnlySet<string>> PushAsync(string accessToken, JsonElement operations)
    {
        if (operations.GetArrayLength() == 0) return new HashSet<string>();
        var result = await SendAsync(HttpMethod.Post, "/api/v1/sync/push",
            new { operations }, accessToken);
        return result.GetProperty("acknowledgements").EnumerateArray()
            .Select(value => value.GetProperty("operationId").GetString()!)
            .ToHashSet(StringComparer.Ordinal);
    }

    public async Task<PullResult> PullAsync(string accessToken, long cursor)
    {
        var result = await SendAsync(HttpMethod.Get,
            $"/api/v1/sync/pull?cursor={cursor}&limit=500", null, accessToken);
        var operations = result.GetProperty("operations").EnumerateArray().Select(value =>
            new PulledOperation(
                value.GetProperty("operationId").GetString()!, value.GetProperty("serverSeq").GetInt64(),
                value.GetProperty("deviceId").GetString()!, value.GetProperty("deviceCounter").GetInt64(),
                value.GetProperty("entityType").GetString()!, value.GetProperty("entityId").GetString()!,
                value.GetProperty("action").GetString()!, value.GetProperty("changedFields").Clone(),
                value.GetProperty("occurredAt").GetInt64())).ToArray();
        return new PullResult(result.GetProperty("cursor").GetInt64(), operations);
    }

    public async Task UploadMediaAsync(string accessToken, IReadOnlyList<SyncMediaObject> media)
    {
        foreach (var batch in media.Chunk(100))
        {
            var response = await SendAsync(HttpMethod.Post, "/api/v1/media/prepare", new {
                objects = batch.Select(value => new { sha256 = value.Sha256, byteCount = value.ByteCount,
                    mimeType = value.MimeType }),
            }, accessToken);
            var results = response.GetProperty("objects").EnumerateArray().ToArray();
            for (var index = 0; index < results.Length; index++)
            {
                if (results[index].GetProperty("present").GetBoolean()) continue;
                var uploadId = results[index].GetProperty("uploadId").GetString()!;
                await using var input = File.OpenRead(batch[index].FilePath); var part = 0;
                while (input.Position < input.Length)
                {
                    var bytes = new byte[checked((int)Math.Min(5 * 1024 * 1024L, input.Length - input.Position))];
                    await input.ReadExactlyAsync(bytes);
                    using var request = new HttpRequestMessage(HttpMethod.Put,
                        $"/api/v1/media/{uploadId}/chunk/{part++}");
                    request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
                    request.Content = new ByteArrayContent(bytes);
                    request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
                    using var chunkResponse = await http.SendAsync(request);
                    if (chunkResponse.StatusCode != System.Net.HttpStatusCode.NoContent)
                        throw new InvalidOperationException($"媒体分片上传失败：HTTP {(int)chunkResponse.StatusCode}");
                }
                await SendAsync(HttpMethod.Post, $"/api/v1/media/{uploadId}/complete", new { }, accessToken);
            }
        }
    }

    public async Task<byte[]> DownloadMediaAsync(string accessToken, string sha256)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, $"/api/v1/media/{sha256}");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
        using var response = await http.SendAsync(request);
        if (!response.IsSuccessStatusCode)
            throw new InvalidOperationException($"媒体下载返回 HTTP {(int)response.StatusCode}");
        return await response.Content.ReadAsByteArrayAsync();
    }

    private static AuthSession Session(JsonElement value)
    {
        var access = value.GetProperty("accessToken").GetString()!;
        var payload = access.Split('.').ElementAtOrDefault(1)
            ?? throw new InvalidDataException("服务端返回了无效 access token");
        payload = payload.Replace('-', '+').Replace('_', '/').PadRight((payload.Length + 3) / 4 * 4, '=');
        using var claims = JsonDocument.Parse(Convert.FromBase64String(payload));
        var accountId = value.TryGetProperty("accountId", out var account)
            ? account.GetString() : claims.RootElement.TryGetProperty("sub", out var subject)
                ? subject.GetString() : null;
        if (string.IsNullOrWhiteSpace(accountId)) throw new InvalidDataException("access token 缺少账号标识");
        return new AuthSession(accountId, value.GetProperty("workspaceId").GetString()!, access,
            DateTimeOffset.UtcNow.ToUnixTimeSeconds() + value.GetProperty("expiresIn").GetInt64(),
            value.GetProperty("refreshToken").GetString()!);
    }

    private async Task<JsonElement> SendAsync(
        HttpMethod method, string path, object? body = null, string? token = null,
        IReadOnlyCollection<System.Net.HttpStatusCode>? expected = null)
    {
        using var request = new HttpRequestMessage(method, path);
        request.Headers.Accept.ParseAdd("application/json, application/problem+json");
        if (token is not null) request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        if (body is not null && method != HttpMethod.Get)
            request.Content = new StringContent(JsonSerializer.Serialize(body, JsonOptions), Encoding.UTF8, "application/json");
        using var response = await http.SendAsync(request);
        var text = await response.Content.ReadAsStringAsync();
        var accepted = expected?.Contains(response.StatusCode) ?? response.IsSuccessStatusCode;
        if (!accepted)
        {
            var title = "";
            try { title = JsonDocument.Parse(text).RootElement.GetProperty("title").GetString() ?? ""; }
            catch (JsonException) { }
            throw new InvalidOperationException(string.IsNullOrWhiteSpace(title)
                ? $"同步服务返回 HTTP {(int)response.StatusCode}" : title);
        }
        if (string.IsNullOrWhiteSpace(text)) return JsonSerializer.SerializeToElement(new { });
        using var document = JsonDocument.Parse(text);
        return document.RootElement.Clone();
    }

    public void Dispose() => http.Dispose();
}

File: apps/android/app/src/main/java/cn/reviewfault/app/sync/SyncClient.kt
package cn.reviewfault.app.sync

import android.util.Base64
import cn.reviewfault.app.data.SyncMediaObject
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.time.Instant
import org.json.JSONArray
import org.json.JSONObject

data class AuthSession(
    val accountId: String,
    val workspaceId: String,
    val accessToken: String,
    val accessExpiresAt: Long,
    val refreshToken: String,
)

data class PulledOperation(
    val operationId: String,
    val serverSeq: Long,
    val deviceId: String,
    val deviceCounter: Long,
    val entityType: String,
    val entityId: String,
    val action: String,
    val changedFields: JSONObject,
    val occurredAt: Long,
)

data class PullResult(val cursor: Long, val operations: List<PulledOperation>)

class SyncClient(endpoint: String) {
    private val baseUrl = endpoint.trim().trimEnd('/').also { value ->
        val uri = runCatching { URI(value) }.getOrNull()
        require(uri != null && uri.isAbsolute && uri.userInfo == null && uri.query == null &&
            uri.fragment == null && (uri.path.isNullOrEmpty() || uri.path == "/") &&
            (uri.scheme.equals("https", ignoreCase = true) ||
              (uri.scheme.equals("http", ignoreCase = true) &&
               uri.host in setOf("10.0.2.2", "localhost", "127.0.0.1")))) {
            "同步地址必须使用 HTTPS（Android 模拟器本机地址除外）"
        }
    }

    fun register(email: String, password: String, invitationCode: String) {
        request("POST", "/api/v1/auth/register", JSONObject().apply {
            put("email", email.trim()); put("password", password)
            put("invitationCode", invitationCode.trim())
        }, expected = setOf(202))
    }

    fun login(email: String, password: String, deviceId: String, deviceName: String): AuthSession {
        val response = request("POST", "/api/v1/auth/login", JSONObject().apply {
            put("email", email.trim()); put("password", password)
            put("deviceId", deviceId); put("deviceName", deviceName)
        })
        return response.toSession()
    }

    fun refresh(deviceId: String, refreshToken: String): AuthSession {
        val response = request("POST", "/api/v1/auth/refresh", JSONObject().apply {
            put("deviceId", deviceId); put("refreshToken", refreshToken)
        })
        return response.toSession()
    }

    fun logout(accessToken: String) {
        request("POST", "/api/v1/auth/logout", JSONObject(), accessToken, setOf(204))
    }

    fun push(accessToken: String, operations: JSONArray): Set<String> {
        if (operations.length() == 0) return emptySet()
        val response = request("POST", "/api/v1/sync/push", JSONObject().put("operations", operations), accessToken)
        val acknowledgements = response.getJSONArray("acknowledgements")
        return buildSet {
            for (index in 0 until acknowledgements.length()) {
                add(acknowledgements.getJSONObject(index).getString("operationId"))
            }
        }
    }

    fun pull(accessToken: String, cursor: Long): PullResult {
        val response = request("GET", "/api/v1/sync/pull?cursor=$cursor&limit=500", token = accessToken)
        val values = response.getJSONArray("operations")
        val operations = buildList {
            for (index in 0 until values.length()) values.getJSONObject(index).let { value ->
                add(PulledOperation(
                    value.getString("operationId"), value.getLong("serverSeq"),
                    value.getString("deviceId"), value.getLong("deviceCounter"),
                    value.getString("entityType"), value.getString("entityId"),
                    value.getString("action"), value.getJSONObject("changedFields"),
                    value.getLong("occurredAt"),
                ))
            }
        }
        return PullResult(response.getLong("cursor"), operations)
    }

    fun uploadMedia(accessToken: String, media: List<SyncMediaObject>) {
        media.chunked(100).forEach { batch ->
            val response = request("POST", "/api/v1/media/prepare", JSONObject().put("objects",
                JSONArray(batch.map { value -> JSONObject().apply {
                    put("sha256", value.sha256); put("byteCount", value.byteCount); put("mimeType", value.mimeType)
                } })), accessToken)
            val results = response.getJSONArray("objects")
            for (index in 0 until results.length()) {
                val result = results.getJSONObject(index)
                if (result.getBoolean("present")) continue
                val uploadId = result.getString("uploadId"); val value = batch[index]
                value.file.inputStream().use { input ->
                    var part = 0
                    while (true) {
                        val buffer = ByteArray(5 * 1024 * 1024)
                        var count = 0
                        while (count < buffer.size) {
                            val read = input.read(buffer, count, buffer.size - count)
                            if (read < 0) break
                            count += read
                        }
                        val bytes = if (count == buffer.size) buffer else buffer.copyOf(count)
                        if (bytes.isEmpty()) break
                        uploadChunk(accessToken, uploadId, part++, bytes)
                    }
                }
                request("POST", "/api/v1/media/$uploadId/complete", JSONObject(), accessToken)
            }
        }
    }

    fun downloadMedia(accessToken: String, sha256: String): ByteArray {
        val connection = (URL("$baseUrl/api/v1/media/$sha256").openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"; connectTimeout = 15_000; readTimeout = 60_000
            setRequestProperty("Authorization", "Bearer $accessToken")
        }
        try {
            val code = connection.responseCode
            if (code != 200) error("媒体下载返回 HTTP $code")
            return connection.inputStream.use { it.readBytes() }
        } finally { connection.disconnect() }
    }

    private fun uploadChunk(accessToken: String, uploadId: String, index: Int, bytes: ByteArray) {
        val connection = (URL("$baseUrl/api/v1/media/$uploadId/chunk/$index").openConnection() as HttpURLConnection).apply {
            requestMethod = "PUT"; connectTimeout = 15_000; readTimeout = 60_000; doOutput = true
            setRequestProperty("Authorization", "Bearer $accessToken")
            setRequestProperty("Content-Type", "application/octet-stream")
            setFixedLengthStreamingMode(bytes.size); outputStream.use { it.write(bytes) }
        }
        try { if (connection.responseCode != 204) error("媒体分片上传失败：HTTP ${connection.responseCode}") }
        finally { connection.disconnect() }
    }

    private fun JSONObject.toSession(): AuthSession {
        val access = getString("accessToken")
        val payload = access.split('.').getOrNull(1) ?: error("服务端返回了无效 access token")
        val decoded = Base64.decode(payload, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
        val claims = JSONObject(String(decoded))
        val accountId = optString("accountId").ifBlank { claims.optString("sub") }.ifBlank {
            claims.optString("http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier")
        }
        require(accountId.isNotBlank()) { "access token 缺少账号标识" }
        return AuthSession(
            accountId, getString("workspaceId"), access,
            Instant.now().epochSecond + getLong("expiresIn"), getString("refreshToken"),
        )
    }

    private fun request(
        method: String,
        path: String,
        body: JSONObject? = null,
        token: String? = null,
        expected: Set<Int> = setOf(200),
    ): JSONObject {
        val connection = (URL(baseUrl + path).openConnection() as HttpURLConnection).apply {
            requestMethod = method; connectTimeout = 15_000; readTimeout = 30_000
            setRequestProperty("Accept", "application/json, application/problem+json")
            if (token != null) setRequestProperty("Authorization", "Bearer $token")
            if (body != null && method != "GET") {
                doOutput = true; setRequestProperty("Content-Type", "application/json")
                outputStream.use { it.write(body.toString().toByteArray()) }
            }
        }
        try {
            val code = connection.responseCode
            val stream = if (code in 200..299) connection.inputStream else connection.errorStream
            val text = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
            if (code !in expected) {
                val problem = runCatching { JSONObject(text).optString("title") }.getOrNull()
                error(problem?.takeIf(String::isNotBlank) ?: "同步服务返回 HTTP $code")
            }
            return if (text.isBlank()) JSONObject() else JSONObject(text)
        } finally { connection.disconnect() }
    }
}
